theories/analysis/RealFLub.ec

```diff
@@ -7,19 +7,75 @@ pragma +implicits.
 (* lub for functions *)
 op flub (F : 'a -> real) : real = lub (fun r => exists a, F a = r).
 
-lemma flub_upper_bound r (F : 'a -> real) x : 
-    (forall x, F x <= r) => F x <= flub F.
+op is_flub (f: 'a -> real) r =
+  forall x, f x <= r.
+
+op has_flub (f: 'a -> real) =
+  exists r, is_flub f r.
+
+lemma has_flub_lub (f: 'a -> real) :
+  has_flub f <=> has_lub (fun r => exists a, f a = r).
```

Comment:
"If it's upper-bounded and non-empty, it has a `lub`"? That sounds like the definition `op has_lub (E : real -> bool) : bool = nonempty E /\ has_ub E.` which is what I used.

The `split` on line 20 splits my ``has_lub` into a non-empty subgoal and an upper-bounded subgoal.

Reply:
Or which lemma is it that you have in mind? I looked at `has_lub` and just couldn't find anything else especially useful...

Reply:
Although I did find that the second part of the proof (`has_lub (fun r => exists a, f a = r) => has_flub f`) can be simplified significantly, now that I'm looking at it again

Reply:
Oh, interesting. Perhaps `has_flub` is the right name then.

I'll need to dive down a bit in related theories to see if there's anything that nails down a least upper bound as a value, rather than just the fact that it exists.

Reply:
Hmmmm maybe I should change the name back then?